lib/compat/wordpress-6.1/block-template-utils.php

```diff
@@ -296,3 +296,63 @@ function gutenberg_build_block_template_result_from_post( $post ) {
 	}
 	return $template;
 }
+
+/**
+ * Helper function to get the Template Hierarchy for a given slug.
+ * We need to Handle special cases here like `front-page`, `singular` and `archive` templates.
+ *
+ * Noting that we always add `index` as the last fallback template.
+ *
+ * @param string  $slug            The template slug to be created.
+ * @param boolean $is_custom       Indicates if a template is custom or part of the template hierarchy.
+ * @param string  $template_prefix The template prefix for the created template. This is used to extract the main template type ex. in `taxonomy-books` we extract the `taxonomy`.
+ *
+ * @return array<string> The template hierarchy.
+ */
+function get_template_hierarchy( $slug, $is_custom = false, $template_prefix = '' ) {
```

Comment:
One concern with moving this logic to the server is accidentally committing to supporting a function forever. What do you think of:

```suggestion
 * @access private
 * @internal
 */
function get_template_hierarchy( $slug, $is_custom = false, $template_prefix = '' ) {
```

```suggestion
 */
function _gutenberg_get_template_hierarchy( $slug, $is_custom = false, $template_prefix = '' ) {
```

Reply:
I don't think we should have a `gutenberg` prefix in core, also the logic in this function would need to change if the template hierarchy logic changes, and I don't imagine this will happen any time soon.

Reply:
I wish we had something like https://github.com/Brain-WP/Hierarchy in the core.